README.md
# get_stream_page

A reactive page in GetX from a stream without the need to create a controller.

## Introduction

Would you like to have a reactive page with just your stream? Built with GetX, this widget offers 
this facility. You can add reactive resettables to rebuild your body as well. Like a change in auth. 
Errors, connectivity and standby widgets are configured in default mode. Change them if you wish. 
Check the example.

## Tips

The function [obxWidgetBuilder] is inside an Obx. Place reactive verables into it.

##### ✳️ There are two ways to add reactive variables.
1 ) Boot your controller into a StatefulWidget.<p>
1.2 - Pass the reactive variable get inside this function.
-----
2 ) Add the parameters to this list. It doesn't have to be a StatefulWidget.<p>
2.2 - Collect the reactive variable with .value => (So much for Rx or RxList).
-----
1 ) => for StatefulWidget
<p>
2 ) => for StatelessWidget, StatefulWidget 
or pass GetStreamPage directly.

#### Example
- Using the pages with details. <p>
[Full Example](https://api.pub.dev/packages/get_stream_page/example) for more details.

------
Any change in the variables will reassemble the body of your page. As in the example below. In the 
case below, there would be 03 ways to reconstruct the screen: the stream flow, rxAuth and rxList changes.
------
#### GetStreamPage

There is already a Scaffold waiting for the parameters. Quick example.
```dart
class TestGetStreamPage extends StatefulWidget {
  @override
  _TestGetStreamPageState createState() => _TestGetStreamPageState();
}

class _TestGetStreamPageState extends State<TestGetStreamPage> {
  Test2Controller controll_1;

  @override
  void initState() {
    /// ## ✳️ There are two ways to add reactive variables.
    ///-------------------------------------------------------------------
    /// ✅ 1) Boot your controller into a StatefulWidget.
    ///-------------------------------------------------------------------
    controll_1 = Get.find<Test2Controller>();
    super.initState();

    Future.delayed(const Duration(seconds: 10), () {
      ///------------------------------------------
      /// Test to check the reactivity of the screen.
      ///------------------------------------------
      /// 1) 👇🏼
      controll_1.changeAuth = true;
    });

    Future.delayed(const Duration(seconds: 15), () {
      ///------------------------------------------
      /// Test to check the reactivity of the screen.
      ///------------------------------------------
      /// 2) 👇🏼
      Get.find<Test2Controller>().rxList.addAll(dataListPerson);
    });
  }

  @override
    Widget build(BuildContext context) {
      return GetStreamPage<List<Person>>(
        title: Text(
          'Stream Page',
          style: TextStyle(fontSize: 18),
        ),
        stream: streamListPerson,
        listRx: [
          ///-------------------------------------------------------------------
          /// ✅ 2) Add the parameters to this list. No need
          ///  be a StatefulWidget. You can directly pass GetStreamPage.
          ///-------------------------------------------------------------------
          RxItem.input(Get.find<Test2Controller>().rxAuth, 'auth'),
          RxItem.input(Get.find<Test2Controller>().rxList, 'list_user'),
          RxItem.input(Get.find<Test3Controller>().rx_2, 'inter')
        ],
        widgetBuilder: (context, objesctStream, rxSet) {
          // ☑️ This function is inside an Obx. Place reactive verables into it.
          ///---------------------------------------------------------------
          /// 2.2) Collect the reactive variable with .value => (Rx or RxList)
          ///---------------------------------------------------------------
          /// Examples
          print(' TEST -- ${rxSet.getRx('auth').value.toString()} ');
          print(' TEST -- ${rxSet.getRx('list_user').value.length.toString()} ');
  
          ///-------------------------------------------------------------
          /// 1.2) Or pass the reactive variable get inside this function.
          ///-------------------------------------------------------------
          print(' TEST -- ${controll_1.isAuth.toString()} ');
  
          /// 1)
          if (!controll_1.isAuth) {
            /// Or 2) 👇🏼
            //if (!rxSet.getRx('auth').value) {
            return Center(
              child: Text(
                'Please login.',
                style: TextStyle(fontSize: 22),
              ),
            );
          }
  
          ///------------------------------------------
          /// Build your body from the stream data.
          ///------------------------------------------
          final list = objesctStream;
          if (list.isEmpty) {
            return Center(
                child: Text(
              'NOTHING FOUND',
              style: TextStyle(fontSize: 14),
            ));
          }
          return Column(
            children: [
              Expanded(
                child: ListView.builder(
                  itemCount: list.length,
                  itemBuilder: (_, index) {
                    return Card(
                        margin: const EdgeInsets.symmetric(
                            horizontal: 16, vertical: 4),
                        shape: RoundedRectangleBorder(
                            borderRadius: BorderRadius.circular(4)),
                        child: Padding(
                          padding: const EdgeInsets.all(14.0),
                          child: Row(
                            children: [
                              Expanded(
                                child: Text(
                                  'Name: ${list[index].name}',
                                  style: TextStyle(fontSize: 16),
                                ),
                              ),
                              Expanded(
                                child: Text(
                                  'Age: ${list[index].age.toStringAsFixed(2)}',
                                  style: TextStyle(fontSize: 12),
                                ),
                              )
                            ],
                          ),
                        ));
                  },
                ),
              ),
            ],
          );
        },
      );
    }
}

Stream<List<Person>> streamListPerson = (() async* {
  await Future<void>.delayed(Duration(seconds: 3));
  //yield null;
  yield dataListPerson;
  await Future<void>.delayed(Duration(seconds: 4));
  yield dataListPerson2;
  await Future<void>.delayed(Duration(seconds: 5));
  //throw Exception('Erro voluntario');
  yield dataListPerson3;
})();

class Test2Controller extends GetxController {
  final rxAuth = false.obs;

  set changeAuth(bool value) => rxAuth.value = value;

  get isAuth => rxAuth.value;

  final rxList = <Person>[].obs;
}

class Test3Controller extends GetxController {
  final rx_2 = ''.obs;

  set rx_2(value) => rx_2.value = value;
}

class Person {
  final String name;

  final int age;

  Person({this.name, this.age});

  @override
  String toString() {
    return 'Person{name: $name, age: $age}';
  }
}
```
##### If you want to build your body independently, use ```[GetStreamWidget]```.
[Full Example](https://api.pub.dev/packages/get_stream_page/example) for more details.



CHANGELOG.md
## [0.1.0]
* Added ```[GetStreamWidget]```  to build your body independently.

## [0.0.2]
* The readme with the example is very accessible.

## [0.0.1]
* Initial commit


lib/src/ui/get_stream_widget.dart
import 'dart:async';
import 'package:flutter/material.dart';
import 'package:get/get.dart';
import 'package:get_stream_page/src/controller/get_stream_controller.dart';
import 'package:get_stream_page/src/ui/utils/functions.dart';
import 'infra/rx_get_type.dart';

class GetStreamWidget<T> extends StatefulWidget {
  ///[listRx] If you want to add observable variables so that changes to
  ///them generate changes on the screen, add to the list of [RxItem] as
  ///an example.
  final List<RxItem> listRx;

  ///[stream] Just pass your stream and you will receive [streamObject] which is
  ///snapshot.data to build your page.
  final Stream<T> stream;

  /// [widgetErrorBuilder] Widget built by the Object error returned by the
  /// [stream] error.
  final WidgetsErrorBuilder widgetErrorBuilder;

  ///[obxWidgetBuilder] This function is launched every time we receive
  ///snapshot.data through the stream. To set up your page you receive
  ///the context, the [streamObject] which is snapshot.data and an [RxGetSet]
  ///object to harvest the reactive variable by the tag. You can present
  ///something if the user is not logged in, for example.
  final GetWidgetBuilder<T> obxWidgetBuilder;

  /// Start showing [widgetWaiting] until it shows the first data
  final Widget widgetWaiting;

  const GetStreamWidget(
      {Key key,
      @required this.stream,
      this.widgetErrorBuilder,
      this.listRx = const <RxItem>[],
      this.obxWidgetBuilder,
      this.widgetWaiting})
      : super(key: key);

  @override
  _GetStreamWidgetState<T> createState() => _GetStreamWidgetState<T>();
}

class _GetStreamWidgetState<T> extends State<GetStreamWidget<T>> {
  StreamSubscription<T> _subscription;

  GetStreamController<T> _controller;

  Widget _widgetWaiting;

  RxGetSet _rxSet;

  @override
  void initState() {
    _controller = GetStreamController();
    _rxSet = RxGetSet(widget.listRx);
    super.initState();
    _subscribeStream();
    _buildWidgetsDefault();
  }

  @override
  void dispose() {
    _controller.onClose();
    _unsubscribeStream();
    super.dispose();
  }

  @override
  void didUpdateWidget(covariant GetStreamWidget<T> oldWidget) {
    super.didUpdateWidget(oldWidget);

    /*if (oldWidget.listRx != widget.listRx) {
      _rxSet = RxGetSet(widget.listRx);
    }*/

    if (oldWidget.stream != widget.stream) {
      if (_subscription != null) {
        _unsubscribeStream();

        _controller.afterDisconnected();
      }

      _subscribeStream();
    }
  }

  void _subscribeStream() {
    _subscription = widget.stream.listen((data) {
      if (data == null) {
        _controller.afterError(Exception('It cannot return null. 😢'));
      } else {
        _controller.afterData(data);
      }
    }, onError: (Object error) {
      _controller.afterError(error);
    }, onDone: () {
      _controller.afterDone();
    });

    //_controller.afterConnected();
  }

  void _unsubscribeStream() {
    if (_subscription != null) {
      _subscription.cancel();
      _subscription = null;
    }
  }

  @override
  Widget build(BuildContext context) {
    return Obx(() {
      if (_controller.snapshot.connectionState == ConnectionState.waiting) {
        return _widgetWaiting;
      }

      if (_controller.snapshot.hasError) {
        return buildWidgetError(_controller.snapshot.error);
      }

      return widget.obxWidgetBuilder(
          //context, _controller.snapshot.data, widget.listRx);
          context,
          _controller.snapshot.data,
          _rxSet);
    });
  }

  Widget buildWidgetError(Object error) {
    if (widget.widgetErrorBuilder == null) {
      return Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Icon(
              Icons.error_outline,
              color: Colors.red,
              size: 60,
            ),
            Center(
              child: Padding(
                padding: const EdgeInsets.only(top: 16),
                child: Text(
                  'We found an error.\n'
                  'Error: $error',
                  textAlign: TextAlign.center,
                ),
              ),
            )
          ]);
    } else {
      return widget.widgetErrorBuilder(error);
    }
  }

  void _buildWidgetsDefault() {
    if (widget.widgetWaiting == null) {
      _widgetWaiting = Center(
        child: Column(
          mainAxisSize: MainAxisSize.min,
          mainAxisAlignment: MainAxisAlignment.center,
          children: const [
            SizedBox(
              width: 60,
              height: 60,
              child: CircularProgressIndicator(),
            ),
          ],
        ),
      );
    } else {
      _widgetWaiting = widget.widgetWaiting;
    }
  }
}


lib/src/ui/infra/rx_get_type.dart
import 'package:get/get.dart';

class RxGetSet {
  final List<RxItem> listRx;

  RxGetSet(this.listRx);

  RxItem getRx(String tag) =>
      listRx.where((element) => element.tag == tag).first;

/*@override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is RxGetSet &&
          runtimeType == other.runtimeType &&
          listRx == other.listRx;

  @override
  int get hashCode => listRx.hashCode;*/
}

class RxItem {
  final RxInterface rx;
  final String tag;
  final bool isRxList;

  dynamic get value => !isRxList ? rx.value : (rx as RxList);

  const RxItem._(this.rx, this.tag, this.isRxList)
      : assert(rx != null),
        assert(tag != null);

  // assert(tag != null && tag.isNotEmpty);

  const RxItem.input(RxInterface rx, String tag)
      : this._(rx, tag, rx is RxList);

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is RxItem && runtimeType == other.runtimeType && tag == other.tag;

  @override
  int get hashCode => tag.hashCode;
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:get/get.dart';
import 'package:get_stream_page/get_stream_page.dart';

void main() {
  runApp(MyApp());
}

class MyBinddings extends Bindings {
  @override
  void dependencies() {
    Get.put(Test2Controller());
    Get.put(Test3Controller());
  }
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return GetMaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      initialBinding: MyBinddings(),
      initialRoute: AppPages.INITIAL,
      getPages: AppPages.routes,
    );
  }
}

abstract class Routes {
  static const HOME = '/home';
  static const PAGE_1 = '/page_1';
  static const PAGE_2 = '/page_2';
  static const PAGE_3 = '/page_3';
}

class AppPages {
  static const INITIAL = Routes.HOME;
  static final routes = [
    GetPage(name: Routes.HOME, page: () => HomePage()),

    /// ## ✳️ There are two ways to add reactive variables.

    ///--------------------------------------------
    /// ✅ 1) Add the reactive parameters to a list.
    ///---------------------------------------------
    GetPage(
        name: Routes.PAGE_1,
        page: () {
          changeAuth();
          return TestGetStreamPage();
        }),

    ///-------------------------------------------------------------------
    /// ✅ 2) Boot your controller into a StatefulWidget.
    ///-------------------------------------------------------------------
    GetPage(name: Routes.PAGE_2, page: () => TestGetStreamPageStateful()),
    GetPage(
        name: Routes.PAGE_3,
        page: () {
          changeAuth();
          return TestGetStreamWidget();
        }),
  ];
}

void changeAuth() {
  Future.delayed(const Duration(seconds: 8), () {
    ///------------------------------------------
    /// Test to check the reactivity of the screen.
    ///------------------------------------------
    /// 1) 👇🏼
    Get.find<Test2Controller>().rxList.addAll(dataListPerson);
    if (!Get.find<Test2Controller>().isAuth) {
      Get.find<Test2Controller>().changeAuth = true;
    }
  });
}

// ignore: must_be_immutable
class TestGetStreamPage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return WillPopScope(
      onWillPop: () {
        Get.find<Test2Controller>().changeAuth = false;
        Get.find<Test2Controller>().rxList.clear();
        return Future.value(true);
      },
      child: GetStreamPage<List<Person>>(
        title: Text(
          'Stream Page',
          style: TextStyle(fontSize: 18),
        ),
        stream: streamListPerson,

        ///--------------------------------------------
        /// ✅ 1) Add the reactive parameters to a list.
        ///---------------------------------------------
        listRx: [
          RxItem.input(Get.find<Test2Controller>().rxAuth, 'auth'),
          RxItem.input(Get.find<Test2Controller>().rxList, 'list_user'),
          RxItem.input(Get.find<Test3Controller>().rx_2, 'inter')
        ],
        obxWidgetBuilder: (context, objesctStream, rxSet) {
          ///-----------------------------------------------
          /// ✅ 1.2) Collect the reactive variable with .value
          /// => (Rx or RxList)
          ///-----------------------------------------------
          ///  If they are within the function [obxWidgetBuilder],
          /// when they change, a rebuild of the body will occur.

          print(
              // ignore: lines_longer_than_80_chars
              ' TEST -- ${rxSet.getRx('list_user').value.length.toString()} ');

          if (!rxSet.getRx('auth').value) {
            return Center(
              child: Text(
                'Please login.',
                style: TextStyle(fontSize: 22),
              ),
            );
          }

          ///------------------------------------------
          /// Build your body from the stream data.
          ///------------------------------------------
          final list = objesctStream;
          if (list.isEmpty) {
            return Center(
                child: Text(
              'NOTHING FOUND',
              style: TextStyle(fontSize: 14),
            ));
          }
          return Column(
            children: [
              Expanded(
                child: ListView.builder(
                  itemCount: list.length,
                  itemBuilder: (_, index) {
                    return Card(
                        margin: const EdgeInsets.symmetric(
                            horizontal: 16, vertical: 4),
                        shape: RoundedRectangleBorder(
                            borderRadius: BorderRadius.circular(4)),
                        child: Padding(
                          padding: const EdgeInsets.all(14.0),
                          child: Row(
                            children: [
                              Expanded(
                                child: Text(
                                  'Name: ${list[index].name}',
                                  style: TextStyle(fontSize: 16),
                                ),
                              ),
                              Expanded(
                                child: Text(
                                  // ignore: lines_longer_than_80_chars
                                  'Age: ${list[index].age.toStringAsFixed(2)}',
                                  style: TextStyle(fontSize: 12),
                                ),
                              )
                            ],
                          ),
                        ));
                  },
                ),
              ),
            ],
          );
        },
      ),
    );
  }

  Stream<List<Person>> streamListPerson = (() async* {
    await Future<void>.delayed(Duration(seconds: 3));
    //yield null;
    yield dataListPerson;
    await Future<void>.delayed(Duration(seconds: 4));
    yield dataListPerson2;
    await Future<void>.delayed(Duration(seconds: 5));
    //throw Exception('Erro voluntario');
    yield dataListPerson3;
  })();
}

class TestGetStreamPageStateful extends StatefulWidget {
  @override
  _TestGetStreamPageStatefulState createState() =>
      _TestGetStreamPageStatefulState();
}

class _TestGetStreamPageStatefulState extends State<TestGetStreamPageStateful> {
  Test2Controller controll_1;

  @override
  void initState() {
    ///-------------------------------------------------------------------
    /// ✅ 2) Boot your controller into a StatefulWidget.
    ///-------------------------------------------------------------------
    controll_1 = Get.find<Test2Controller>();
    super.initState();

    Future.delayed(const Duration(seconds: 8), () {
      ///------------------------------------------
      /// Test to check the reactivity of the screen.
      ///------------------------------------------
      /// 2) 👇🏼
      controll_1.changeAuth = true;
    });
  }

  @override
  Widget build(BuildContext context) {
    return GetStreamPage<List<Person>>(
      title: Text(
        'Stream Page',
        style: TextStyle(fontSize: 18),
      ),
      stream: streamListPerson,
      obxWidgetBuilder: (context, objesctStream, _) {
        // ☑️ This function is inside an Obx. Place reactive verables into it.

        ///-------------------------------------------------------------
        /// 2.2) Or pass the reactive variable get inside this function.
        ///-------------------------------------------------------------
        ///

        if (!controll_1.isAuth) {
          return Center(
            child: Text(
              'Please login.',
              style: TextStyle(fontSize: 22),
            ),
          );
        }

        ///------------------------------------------
        /// Build your body from the stream data.
        ///------------------------------------------
        final list = objesctStream;
        if (list.isEmpty) {
          return Center(
              child: Text(
            'NOTHING FOUND',
            style: TextStyle(fontSize: 14),
          ));
        }
        return Column(
          children: [
            Expanded(
              child: ListView.builder(
                itemCount: list.length,
                itemBuilder: (_, index) {
                  return Card(
                      margin: const EdgeInsets.symmetric(
                          horizontal: 16, vertical: 4),
                      shape: RoundedRectangleBorder(
                          borderRadius: BorderRadius.circular(4)),
                      child: Padding(
                        padding: const EdgeInsets.all(14.0),
                        child: Row(
                          children: [
                            Expanded(
                              child: Text(
                                'Name: ${list[index].name}',
                                style: TextStyle(fontSize: 16),
                              ),
                            ),
                            Expanded(
                              child: Text(
                                'Age: ${list[index].age.toStringAsFixed(2)}',
                                style: TextStyle(fontSize: 12),
                              ),
                            )
                          ],
                        ),
                      ));
                },
              ),
            ),
          ],
        );
      },
    );
  }

  Stream<List<Person>> streamListPerson = (() async* {
    await Future<void>.delayed(Duration(seconds: 3));
    //yield null;
    yield dataListPerson;
    await Future<void>.delayed(Duration(seconds: 4));
    yield dataListPerson2;
    await Future<void>.delayed(Duration(seconds: 5));
    //throw Exception('Erro voluntario');
    yield dataListPerson3;
  })();

  @override
  void dispose() {
    Get.find<Test2Controller>().changeAuth = false;
    super.dispose();
  }
}

class Test2Controller extends GetxController {
  final rxAuth = false.obs;

  set changeAuth(bool value) => rxAuth.value = value;

  get isAuth => rxAuth.value;

  final rxList = <Person>[].obs;
}

class Test3Controller extends GetxController {
  final rx_2 = ''.obs;

  set rx_2(value) => rx_2.value = value;
}

class HomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('HomeView'),
        centerTitle: true,
      ),
      body: Center(
        child: Column(
          mainAxisSize: MainAxisSize.min,
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            InkWell(
                focusColor: Colors.grey,
                splashColor: Colors.blue,
                onTap: () {
                  Get.toNamed(Routes.PAGE_1);
                },
                child: Text(
                  'Go to the GetStreamPage with reactive variables in a list.',
                  textAlign: TextAlign.center,
                  style: TextStyle(fontSize: 20),
                )),
            const SizedBox(
              height: 20,
            ),
            InkWell(
                focusColor: Colors.grey,
                splashColor: Colors.blue,
                onTap: () {
                  Get.toNamed(Routes.PAGE_2);
                },
                child: Text(
                  'Go to the GetStreamPage with '
                  'reactive variables by controllers.',
                  textAlign: TextAlign.center,
                  style: TextStyle(fontSize: 20),
                )),
            const SizedBox(
              height: 20,
            ),
            InkWell(
                focusColor: Colors.grey,
                splashColor: Colors.blue,
                onTap: () {
                  Get.toNamed(Routes.PAGE_3);
                },
                child: Text(
                  'Go to the GetStreamWidget',
                  textAlign: TextAlign.center,
                  style: TextStyle(fontSize: 20),
                )),
          ],
        ),
      ),
    );
  }
}

// ignore: must_be_immutable
class TestGetStreamWidget extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('GetStreamWidget'),
        centerTitle: true,
      ),
      body: WillPopScope(
        onWillPop: () {
          Get.find<Test2Controller>().changeAuth = false;
          Get.find<Test2Controller>().rxList.clear();
          return Future.value(true);
        },
        child: GetStreamWidget<List<Person>>(
          stream: streamListPerson,
          listRx: [
            RxItem.input(Get.find<Test2Controller>().rxAuth, 'auth'),
            RxItem.input(Get.find<Test2Controller>().rxList, 'list_user'),
            RxItem.input(Get.find<Test3Controller>().rx_2, 'inter')
          ],
          obxWidgetBuilder: (context, objesctStream, rxSet) {
            print(
                // ignore: lines_longer_than_80_chars
                ' TEST -- ${rxSet.getRx('list_user').value.length.toString()} ');

            if (!rxSet.getRx('auth').value) {
              return Center(
                child: Text(
                  'Please login.',
                  style: TextStyle(fontSize: 22),
                ),
              );
            }

            ///------------------------------------------
            /// Build your body from the stream data.
            ///------------------------------------------
            final list = objesctStream;
            if (list.isEmpty) {
              return Center(
                  child: Text(
                'NOTHING FOUND',
                style: TextStyle(fontSize: 14),
              ));
            }
            return Column(
              children: [
                Expanded(
                  child: ListView.builder(
                    itemCount: list.length,
                    itemBuilder: (_, index) {
                      return Card(
                          margin: const EdgeInsets.symmetric(
                              horizontal: 16, vertical: 4),
                          shape: RoundedRectangleBorder(
                              borderRadius: BorderRadius.circular(4)),
                          child: Padding(
                            padding: const EdgeInsets.all(14.0),
                            child: Row(
                              children: [
                                Expanded(
                                  child: Text(
                                    'Name: ${list[index].name}',
                                    style: TextStyle(fontSize: 16),
                                  ),
                                ),
                                Expanded(
                                  child: Text(
                                    // ignore: lines_longer_than_80_chars
                                    'Age: ${list[index].age.toStringAsFixed(2)}',
                                    style: TextStyle(fontSize: 12),
                                  ),
                                )
                              ],
                            ),
                          ));
                    },
                  ),
                ),
              ],
            );
          },
        ),
      ),
    );
  }

  Stream<List<Person>> streamListPerson = (() async* {
    await Future<void>.delayed(Duration(seconds: 3));
    //yield null;
    yield dataListPerson;
    await Future<void>.delayed(Duration(seconds: 4));
    yield dataListPerson2;
    await Future<void>.delayed(Duration(seconds: 5));
    //throw Exception('Erro voluntario');
    yield dataListPerson3;
  })();
}

final dataListPerson = <Person>[
  Person(name: 'Rafaela Pinho', age: 30),
  Person(name: 'Paulo Emilio Silva', age: 45),
  Person(name: 'Pedro Gomes', age: 18),
  Person(name: 'Orlando Guerra', age: 23),
  Person(name: 'Zacarias Triste', age: 15),
];

final dataListPerson2 = <Person>[
  Person(name: 'Rafaela Pinho', age: 30),
  Person(name: 'Paulo Emilio Silva', age: 45),
  Person(name: 'Pedro Gomes', age: 18),
  Person(name: 'Orlando Guerra', age: 23),
  Person(name: 'Zacarias Triste', age: 15),
  Person(name: 'Antonio Rabelo', age: 33),
  Person(name: 'Leticia Maciel', age: 47),
  Person(name: 'Patricia Oliveira', age: 19),
  Person(name: 'Pedro Lima', age: 15),
  Person(name: 'Junior Rabelo', age: 33),
  Person(name: 'Lucia Maciel', age: 47),
  Person(name: 'Ana Oliveira', age: 19),
  Person(name: 'Thiago Silva', age: 33),
  Person(name: 'Charles Ristow', age: 47),
  Person(name: 'Raquel Montenegro', age: 19),
  Person(name: 'Rafael Peireira', age: 15),
  Person(name: 'Nome Comum', age: 33),
];

final dataListPerson3 = <Person>[
  Person(name: 'Rafaela Pinho', age: 30),
  Person(name: 'Paulo Emilio Silva', age: 45),
  Person(name: 'Pedro Gomes', age: 18),
  Person(name: 'Orlando Guerra', age: 23),
  Person(name: 'Ana Pereira', age: 23),
  Person(name: 'Zacarias Triste', age: 15),
  Person(name: 'Antonio Rabelo', age: 33),
  Person(name: 'Leticia Maciel', age: 47),
  Person(name: 'Patricia Oliveira', age: 19),
  Person(name: 'Pedro Lima', age: 15),
  Person(name: 'Fabio Melo', age: 51),
  Person(name: 'Junior Rabelo', age: 33),
  Person(name: 'Lucia Maciel', age: 47),
  Person(name: 'Ana Oliveira', age: 19),
  Person(name: 'Thiago Silva', age: 33),
  Person(name: 'Charles Ristow', age: 47),
  Person(name: 'Raquel Montenegro', age: 19),
  Person(name: 'Rafael Peireira', age: 15),
  Person(name: 'Thiago Ferreira', age: 33),
  Person(name: 'Joaquim Gomes', age: 18),
  Person(name: 'Esther Guerra', age: 23),
  Person(name: 'Pedro Braga', age: 19),
  Person(name: 'Milu Silva', age: 17),
  Person(name: 'William Ristow', age: 47),
  Person(name: 'Elias Tato', age: 22),
  Person(name: 'Dada Istomesmo', age: 44),
  Person(name: 'Nome Incomum', age: 52),
  Person(name: 'Qualquer Nome', age: 9),
  Person(name: 'First Last', age: 11),
  Person(name: 'Bom Dia', age: 23),
  Person(name: 'Bem Mequiz', age: 13),
  Person(name: 'Mal Mequer', age: 71),
  Person(name: 'Quem Sabe', age: 35),
  Person(name: 'Miriam Leitao', age: 33),
  Person(name: 'Gabriel Mentiroso', age: 19),
  Person(name: 'Caio Petro', age: 27),
  Person(name: 'Tanto Nome', age: 66),
  Person(name: 'Nao Diga', age: 33),
  Person(name: 'Fique Queto', age: 11),
  Person(name: 'Cicero Gome', age: 37),
  Person(name: 'Carlos Gome', age: 48),
  Person(name: 'Mae Querida', age: 45),
  Person(name: 'Exausto Nome', age: 81),
];

class Person {
  final String name;

  final int age;

  Person({this.name, this.age});

  @override
  String toString() {
    return 'Person{name: $name, age: $age}';
  }
}
